## Homework
The goal of this homework is to get familiar with MLflow.

### Q1. Install MLflow
To get started with MLflow you'll need to install the MLflow Python package.

For this we recommend creating a separate Python environment, for example, you can use conda environments, and then install the package there with pip or conda.

Once you installed the package, run the command mlflow --version and check the output.

What's the version that you have?

In [1]:
!mlflow --version

mlflow, version 1.28.0


Answer is `1.28.0`

### Q2. Download and preprocess the data

How many files were saved to OUTPUT_FOLDER?

In [5]:
!python ./steps/preprocess_data.py --raw_data_path ../01-intro/data --dest_path ./output

Answer is `4`

### Q3. Train a model with autolog

How many parameters are automatically logged by MLflow?

In [7]:
!python ./steps/train.py

2022/08/21 16:16:57 INFO mlflow.tracking.fluent: Experiment with name 'my-homework-1' does not exist. Creating a new experiment.
2022/08/21 16:16:57 INFO mlflow.tracking.fluent: Autologging successfully enabled for sklearn.
2022/08/21 16:17:14 WARNING mlflow.utils.autologging_utils: MLflow autologging encountered a warning: "/usr/local/anaconda3/envs/mlops-zoomcamp-course/lib/python3.9/site-packages/_distutils_hack/__init__.py:30: UserWarning: Setuptools is replacing distutils."


Answer is `17`

### Q4. Launch the tracking server locally

In addition to `backend-store-uri`, what else do you need to pass to properly configure the server?

Answer is `default-artifact-root`

### Q5. Tune the hyperparameters of the model

What's the best validation RMSE that you got?

In [14]:
!python ./steps/hpo.py

100%|█████████| 50/50 [05:11<00:00,  6.23s/trial, best loss: 6.6284257482044735]


Answer is `6.628`

### Q6. Promote the best model to the model registry

What is the test RMSE of the best model?

In [31]:
!python ./steps/register_model.py

2022/08/21 17:06:56 INFO mlflow.tracking.fluent: Experiment with name 'random-forest-best-models' does not exist. Creating a new experiment.
2022/08/21 17:07:10 WARNING mlflow.utils.autologging_utils: MLflow autologging encountered a warning: "/usr/local/anaconda3/envs/mlops-zoomcamp-course/lib/python3.9/site-packages/_distutils_hack/__init__.py:30: UserWarning: Setuptools is replacing distutils."
Successfully registered model 'random-forest-best-model'.
2022/08/21 17:08:08 INFO mlflow.tracking._model_registry.client: Waiting up to 300 seconds for model version to finish creation.                     Model name: random-forest-best-model, version 1
Created version '1' of model 'random-forest-best-model'.


Answer is `6.548` for me. The closest is `6.55`